packages/rest/src/router/external-express-routes.ts

```diff
@@ -0,0 +1,176 @@
+    spec: RouterSpec = {paths: {}},
+  ) {
+    this._externalRoutes.use(basePath, router);
+
+    spec = rebaseOpenApiSpec(spec, basePath);
+    assignRouterSpec(this._specForExternalRoutes, spec);
+  }
+
+  find(request: Request): ResolvedRoute {
+    // TODO(bajtos) look up OpenAPI spec for matching LB3 route?
+    return new ExternalRoute(
+      this._externalRoutes,
+      this._staticRoutes,
+      request.method,
+      request.url,
+      // TODO(bajtos) look-up the spec from `this.specForExternalRoutes`
+      {
+        description: 'LoopBack static assets route',
+        'x-visibility': 'undocumented',
+        responses: {},
+      },
+    );
+  }
+}
+
+class ExternalRoute implements RouteEntry, ResolvedRoute {
+  // ResolvedRoute API
+  readonly pathParams: PathParameterValues = [];
+  readonly schemas: SchemasObject = {};
+
+  constructor(
+    private readonly _externalRouter: express.Router,
+    private readonly _staticAssets: express.Router,
+    public readonly verb: string,
+    public readonly path: string,
+    public readonly spec: OperationObject,
+  ) {}
+
+  updateBindings(requestContext: Context): void {
+    // no-op
+  }
+
+  async invokeHandler(
+    context: RequestContext,
+    args: OperationArgs,
+  ): Promise<OperationRetval> {
+    const {request, response} = context;
+    let handled = await executeRequestHandler(
+      this._externalRouter,
+      request,
+      response,
+    );
+    if (handled) return;
+
+    handled = await executeRequestHandler(
+      this._staticAssets,
+      request,
+      response,
+    );
+    if (handled) return;
+
+    // Express router called next, which means no route was matched
+    throw new HttpErrors.NotFound(
+      `Endpoint "${request.method} ${request.path}" not found.`,
+    );
+  }
+
+  describe(): string {
+    // TODO(bajtos) provide better description for Express routes with spec
+    return `external Express route "${this.verb} ${this.path}"`;
+  }
+}
+
+export function rebaseOpenApiSpec<T extends Partial<OpenApiSpec>>(
+  spec: T,
+  basePath: string,
+): T {
+  if (!spec.paths) return spec;
+  if (!basePath || basePath === '/') return spec;
+
+  const localPaths = spec.paths;
+  // Don't modify the spec object provided to us.
+  spec = Object.assign({}, spec);
+  spec.paths = {};
+  for (const url in localPaths) {
+    spec.paths[`${basePath}${url}`] = localPaths[url];
+  }
+
+  return spec;
+}
+
+/**
+ * Execute an Express-style callback-based request handler.
+ *
+ * @param handler
+ * @param request
+ * @param response
+ * @returns A promise resolved to:
+ * - `true` when the request was handled
+ * - `false` when the handler called `next()` to proceed to the next
+ *    handler (middleware) in the chain.
+ */
+export function executeRequestHandler(
+  handler: express.RequestHandler,
+  request: Request,
+  response: Response,
+): Promise<boolean> {
+  return new Promise((resolve, reject) => {
+    const onceFinished = () => resolve(true);
```

Comment:
Let's use `on-finished` package.

Reply:
+1

This code was copied from StaticAssetsRoute as-is.

Reply:
See https://github.com/strongloop/loopback-next/pull/2374